### PR TITLE
CDAP-3624 Converting lineage request time period into scan range

### PR DESCRIPTION
**Pull request by poornachandra**
Scan range is based on program start times.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java [hunk 1]
```diff
@@ -482,4 +484,9 @@ WorkflowStatistics getWorkflowStatistics(Id.Workflow workflowId, long startTime,
    */
   Collection<WorkflowDataset.WorkflowRunRecord> retrieveSpacedRecords(Id.Workflow workflow, String runId,
                                                                        int limit, long timeInterval);
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: nit: indentation

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java [hunk 2]
```diff
@@ -482,4 +484,9 @@ WorkflowStatistics getWorkflowStatistics(Id.Workflow workflowId, long startTime,
...
+
+  /**
+   * @return programs that were running between given start and end time.
+   */
+  Set<RunId> getRunningInRange(final long startTimeInSecs, final long endTimeInSecs);
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: why are params final?

**Comment by poornachandra** [comment]
Build - http://builds.cask.co/browse/CDAP-RBT425-7

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageGenerator.java [hunk 3]
```diff
@@ -125,15 +151,75 @@ private Lineage doComputeLineage(final Id.NamespacedId sourceData, long start, l
...
+    return new ScanRange(earliest, latest + 1, new Predicate<Relation>() {
+      @Override
+      public boolean apply(Relation input) {
+        return runIds.contains(input.getRun());
+      }
+    });
+  }
+
+  @VisibleForTesting
+  static class ScanRange {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageGenerator.java]: nit: ``ScanRangeWithFilter``?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageGenerator.java [hunk 4]
```diff
@@ -49,40 +59,54 @@
...
-  private Lineage doComputeLineage(final Id.NamespacedId sourceData, long start, long end, int levels) {
-    LOG.trace("Computing lineage for data {}, start {}, end {}, levels {}", sourceData, start, end, levels);
+  private Lineage doComputeLineage(final Id.NamespacedId sourceData, long startMillis, long endMillis, int levels) {
+    LOG.trace("Computing lineage for data {}, startMillis {}, endMillis {}, levels {}",
+              sourceData, startMillis, endMillis, levels);
+
+    // Convert start time and end time period into scan keys in terms of program start times.
+    Set<RunId> runningInRange = store.getRunningInRange(TimeUnit.MILLISECONDS.toSeconds(startMillis),
+                                                        TimeUnit.MILLISECONDS.toSeconds(endMillis));
+    if (LOG.isTraceEnabled()) {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageGenerator.java]: why is this required ?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java [hunk 5]
```diff
@@ -946,4 +949,72 @@ public int hashCode() {
...
+    writeStartRecord(run1);
+    writeStartRecord(run2);
+    writeStartRecord(run3);
+    writeStartRecord(run4);
+    writeStartRecord(run5);
+    writeStartRecord(run6);
+
+    Assert.assertEquals(runsToTime(run1, run2), runIdsToTime(store.getRunningInRange(1, 30)));
+    Assert.assertEquals(runsToTime(run1, run2, run5, run3), runIdsToTime(store.getRunningInRange(30, 50)));
+    Assert.assertEquals(runsToTime(run1, run2, run3, run4, run5, run5, run6),
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java]: why is run5 found two times?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java [hunk 6]
```diff
@@ -946,4 +949,72 @@ public int hashCode() {
...
+                        runIdsToTime(store.getRunningInRange(1, 71)));
+    Assert.assertEquals(runsToTime(run1, run2, run3, run4, run5, run5, run6),
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java]: ditto

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java [hunk 7]
```diff
@@ -946,4 +949,72 @@ public int hashCode() {
...
+                        runIdsToTime(store.getRunningInRange(50, 71)));
+    Assert.assertEquals(ImmutableSet.of(), runIdsToTime(store.getRunningInRange(1, 10)));
+
+    writeStopRecord(run1, 45000);
+    writeStopRecord(run3, 55000);
+    writeSusspendedRecord(run5);
+
+    Assert.assertEquals(runsToTime(run2, run3, run4, run5, run5, run6),
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/DefaultStoreTest.java]: ditto

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java [hunk 8]
```diff
@@ -332,18 +333,16 @@ private String getRunId(Row row) {
...
   }
 
-  private byte[] getDatasetScanEndKey(Id.DatasetInstance datasetInstance) {
-    // TODO: the scan space can be further reduced by using min(all running program's start time) instead of 0L.
-    long invertedTime = invertTime(0L);
-    MDSKey.Builder builder = new MDSKey.Builder();
-    addDataset(builder, datasetInstance);
-    builder.add(invertedTime);
+  private byte[] getDatasetScanStartKey(Id.DatasetInstance datasetInstance, long start) {
+    return getDatasetScanKey(datasetInstance, start + 1);
```
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java]: can you add a comment about why we are doing +1 here

**Comment by bdmogal** [comment]
@poornachandra only minor comments. LGTM otherwise

**Comment by poornachandra** [comment]
@bdmogal @gsps1 Thanks for the review. Since the comments are minor, can I do it in an upcoming PR?

File: cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageGenerator.java (hunk 4 above)
Comment (poornachandra) [inline cdap-app-fabric/src/main/java/co/cask/cdap/metadata/LineageGenerator.java]: I am printing the size of the container in the log message, so thought it safe to put a check - in case size is not a constant time function.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java (hunk 8 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java]: Good point.

**Comment by bdmogal** [comment]
@poornachandra ok to address comments in a future PR, since that depends on this. LGTM.

**Comment by poornachandra** [comment]
Build is green - http://builds.cask.co/browse/CDAP-RBT425-9
